data_structures/rb_tree2.cpp

```diff
@@ -0,0 +1,1749 @@
+     * @return unsigned int number of children this node has
+     */
+    unsigned int get_num_children(RBTNode* node) {
+        // to check how many children the node has
+        unsigned int num_children = 0;
+        if (node->left != nullptr)
+            num_children++;
+        if (node->right != nullptr)
+            num_children++;
+
+        return num_children;
+    }
+
+    /**
+     * @brief Fixes double-blackness issue caused by trying to remove
+     * a node from tree - to ensure each tree path has an equal number
+     * of black nodes (one of the properties of Red Black Trees).
+     *
+     * @param replacement_node the double-black node
+     */
+    void fix_double_black(RBTNode* replacement_node) {
+        // sibling not needed if replacement node is at the tree's root
+        if (replacement_node == root) {
+            replacement_node->color = COLOR_BLACK;
+        } else {
+            // now using sibling for below conditions
+            RBTNode* sibling = GetSibling(replacement_node);
+
+            // sibling is black - check sibling's kids
+            if (sibling->color == COLOR_BLACK) {
+                // at least one red child
+                if (has_red_child(sibling)) {
+                    // rotate/recolor to fix double-blackness
+                    CheckSiblingConditions(replacement_node, sibling);
+                } else {
+                    // no red children
+
+                    // shift blackness level up to parent
+                    sibling->color = COLOR_RED;
+                    replacement_node->color = COLOR_BLACK;
+                    replacement_node->parent->color++;
+                    // blackness shift may make parent double-black,
+                    // so fix if needed
+                    if (replacement_node->parent->color == COLOR_DOUBLE_BLACK) {
+                        fix_double_black(replacement_node->parent);
+                    }
+                }
+            } else if (sibling->color == COLOR_RED) {
+                // or replacement's sibling is red,
+                // rotate to give replacement a black sibling
+
+                // recoloring needed for rotation
+                sibling->parent->color = COLOR_RED;
+                sibling->color = COLOR_BLACK;
+
+                // rotate right for left sibling
+                if (ParentToMePtr(sibling) == sibling->parent->left) {
+                    RotateRight(sibling);
+                } else if (ParentToMePtr(sibling) == sibling->parent->right) {
+                    // rotate left for right sibling
+                    RotateLeft(sibling);
+                }
+                // try fixing double-blackness again after
+                // giving replacement node a black sibling
+                fix_double_black(replacement_node);
+            }
+        }
+    }
+
+    /**
+     * @brief Checks if passed node has a red child.
+     *
+     * @param n
+     * @return true
```

Comment:
you have to specify when it returns this value 
i.e
```suggestion
     * @return true if the node HAS a red child 
```